### PR TITLE
[irods/irods#8203] Add release notes for 4.3.4 (main)

### PR DESCRIPTION
**Pull request by korydraughn**
This is a WIP.

Please double check issue numbers and wording.

NTS: Check for duplicate issue numbers.

Leaving in draft for now. Feedback welcome.

### CODE REVIEW

File: docs/release_notes.md [hunk 1]
```diff
@@ -1,5 +1,186 @@
...
+
+TODO: issue #7838 is still open, but has commits against it? 
+    - do we have other cases of this for previous releases?
+    - it's also part of the 4.3.5 milestone
+    - i see two paths forward:
+        - rename it to include something like "part 1", add it to 4.3.4 milestone, close it, and create a "part 2" issue for 4.3.5 milestone
+        - or, remove it from the release notes entirely
+        - are there other solutions to this?
+    - are release notes meant to only include "closed" issues?
+[irods/irods#7838,irods/irods#8122] CMake: remove old CentOS 7 support code
```
Comment (korydraughn) [inline docs/release_notes.md:180]: @trel Thoughts on this?

**Comment by korydraughn** [comment]
This was put together based on the commits to irods/irods and irods/irods_client_icommands, not the issue lists or milestones for any repository.

File: docs/release_notes.md [hunk 2]
```diff
@@ -1,5 +1,186 @@
...
+- Deprecate modification of ACL policy (#6843).
+- Deprecate GeneralRowInsert and GeneralRowPurge (#7608).
+- Deprecate update_deprecated_database_columns.py (#7835).
+- Deprecate `imeta qu` (#7959).
+- Deprecate `imeta` interactive mode (#7961).
+- Deprecate server monitoring microservices (#7977).
+- Deprecate DataObjLock and DataObjUnlock (#7979).
+- Deprecate unused global variables in resource manager implementation. (#8042).
+- Deprecate `fillGenQueryInpFromStrCond()` (#8088).
+- Deprecate `forkAndExec()` (#8109).
```
Comment (korydraughn) [inline docs/release_notes.md:134]: Remove parens from functions.

File: docs/release_notes.md [hunk 3]
```diff
@@ -1,5 +1,186 @@
...
+- Rework logic in `_cllExecSqlNoResult` to return proper error codes (#7599, #7440).
+- Fix flex warning for GenQuery2 lexer rules (#7685).
+- Validate and reject invalid zone names (#7722).
+- Return non-zero exit code from `iadmin` on failure (#7734).
+- Remove `name` from exception (#7752).
+- Ignore SIGPIPE so agents do not terminate immediately (#7933, #7934).
+- Change if-statement so ticket use-count is only updated on first pass (#7967).
+- Write client and proxy user info to `ips` data file in correct order (#8001).
+- Modify JSON schema file to enforce required attributes (#8017).
+- Use `std::tolower()` with `std::transform()` correctly (#8045).
```
Comment (korydraughn) [inline docs/release_notes.md:150]: Remove parens from functions.

File: docs/release_notes.md [hunk 4]
```diff
@@ -1,5 +1,186 @@
...
+### Fixed
+
+- Avoid SIGABRT by setting pointers to null after deallocation (#3581).
+- Return errors on one-sided encryption in parallel transfer (#4984).
+- Use `ProcessType` to detect server vs pure client (#6684).
+- Rework logic in `_cllExecSqlNoResult` to return proper error codes (#7599, #7440).
+- Fix flex warning for GenQuery2 lexer rules (#7685).
+- Validate and reject invalid zone names (#7722).
+- Return non-zero exit code from `iadmin` on failure (#7734).
+- Remove `name` from exception (#7752).
```
Comment (korydraughn) [inline docs/release_notes.md:145]: Consider using more words to make this clearer in what it is saying.

File: docs/release_notes.md (hunk 1 above)
Comment (trel) [reply docs/release_notes.md:180]: if all our commits are against an issue (which they should be), and all issues are in a milestone (which they should be), then these notes can track the issues closed.

7838 isn't in this milestone, so it doesn't seem it should be mentioned as a closed issue.

so yes, you could do a part1/part2 thing - is the work in part1 substantial enough to warrant mention in this set of release notes?   if not... then just leave out the mention.    it seems it's only CMakeLists.txt...

Comment (korydraughn) [reply docs/release_notes.md:180]: 7838 isn't substantial. Just can't remember if we've ever encountered this situation, which I'm betting we have.

I'm going to leave it out since it's not in this milestone.

Thinking about this a little more, the housekeeping issues are identical to this situation since the issue for those are never closed and exist in an ongoing milestone.

Comment (korydraughn) [reply docs/release_notes.md]: Done.

File: docs/release_notes.md (hunk 2 above)
Comment (korydraughn) [reply docs/release_notes.md]: Done.

File: docs/release_notes.md (hunk 3 above)
Comment (korydraughn) [reply docs/release_notes.md]: Done.

File: docs/release_notes.md (hunk 4 above)
Comment (korydraughn) [reply docs/release_notes.md]: Done.

File: docs/release_notes.md [hunk 5]
```diff
@@ -1,5 +1,176 @@
...
+
+- Improve documentation (irods/irods_client_icommands#433, #2600, #7622, #7701, #7784, #7952, #8000).
+
+    [irods/irods_client_icommands#433] fix ibun help text
+    [#2600] Doxygen: Add replica_open.h
+    [irods/irods#2600] update irsync help text
+    [irods/irods#7622] Replace examples in iquest help text with link to web page.
+    [#7701,#8000] Add documentation for updating the schema version numbers.
+    [irods/irods#7784] Remove irrelevant help text from iclienthints.
+    [irods/irods#7952] Use root resource instead of leaf resource in ichksum help text.
```
Comment (korydraughn) [inline docs/release_notes.md:28]: Remove indented lines once L20 is confirmed to be good.

File: docs/release_notes.md [hunk 6]
```diff
@@ -1,5 +1,176 @@
...
+    [#7795] iquery: Add test for query tab in where clause
+    [#7961] Update test for imeta deprecation message
+    [#7990] Change test so USER_SOCK_CONNECT_TIMEDOUT is also accepted.
+    [#8046] Connect to server as correct user in rc_switch_user test.
+    [#8047] Guarantee test users are removed if rc_switch_user test fails.
+    [#8134,#8135] GenQuery2: Add test for column mappings.
+    [#8166] Add test for "istream read --count".
+    [#8182] Add test to guard against whitespace issues in GenQuery2.
+    [#8183] Add ASCII printable file creation function for testing.
+    [#8185] Fix memory leaks in GenQuery2 C++ unit test.
```
Comment (korydraughn) [inline docs/release_notes.md:45]: Remove indented lines once L30 is confirmed to be good.

File: docs/release_notes.md [hunk 7]
```diff
@@ -1,5 +1,176 @@
...
+
+- Clean up code (#4984, #5800, #6972).
+
+    [#4984] Clean up headers in rsDataGet and rsDataPut
+    [#5800] catch boost::bad_lexical_cast by reference
+    [#6972] Silence clang-tidy array decay checks for exception macros
```
Comment (korydraughn) [inline docs/release_notes.md:51]: Remove indented lines once L47 is confirmed to be good.

File: docs/release_notes.md [hunk 8]
```diff
@@ -1,5 +1,176 @@
...
+    [#6546] Remove unused `ruleAdmin.cpp`
+    [#7919] control-plane: remove unused boost.atomic include
+    [#7942] Remove irods_assert.hpp
+    [#8010] Remove duplicate function declaration
+    [#8133] Remove duplicate macro definition for GenQuery1.
+    [#8201] `parseCommandLine.cpp`: remove unused win32 compatibility code
+    [#8201] `rodsLog.cpp`: remove unused win32 compatibility code
+    [#8201] `rodsPath.cpp`: remove unused win32 compatibility code
+    [#8201] `sockComm.cpp`: remove unused win32 compatibility code
+    [#8204] Remove unused `irods_zone_info.cpp`
```
Comment (korydraughn) [inline docs/release_notes.md:69]: Remove indented lines once L53 is confirmed to be good.

File: docs/release_notes.md [hunk 9]
```diff
@@ -1,5 +1,176 @@
...
+
+- Replace use of externals-provided libarchive with distro-provided libarchive (#6250, #7286).
+
+    [#6250,#7286] Use system libarchive
+    [irods/irods#7286] Remove another reference to externals-provided libarchive
+    [irods/irods#7286] Remove references to externals-provided libarchive
```
Comment (korydraughn) [inline docs/release_notes.md:75]: Remove indented lines once L71 is confirmed to be good.

File: docs/release_notes.md [hunk 10]
```diff
@@ -1,5 +1,176 @@
...
+
+- Improve CMake (#6319, #6584).
+
+    [#6319] CMake: include headers in target sources
+    [#6584] Make `CMAKE_POSITION_INDEPENDENT_CODE` overridable
```
Comment (korydraughn) [inline docs/release_notes.md:80]: Remove indented lines once L77 is confirmed to be good.

File: docs/release_notes.md [hunk 11]
```diff
@@ -1,5 +1,176 @@
...
+
+- Update clang-format / clang-tidy configuration (#6970, #7313, #7751, #7821, #7822).
+
+    [#6970] clang-format: Allow single-line empty curly braces
+    [#7313] Demote clang-tidy checks for magic numbers to warning.
+    [#7751] clang-tidy: demote `cppcoreguidelines-macro-usage` to warning
+    [#7821] clang-tidy: demote `cppcoreguidelines-avoid-c-arrays` to warning
+    [#7821] clang-tidy: demote `modernize-avoid-c-arrays` to warning
+    [#7822] clang-tidy: demote `cppcoreguidelines-pro-type-vararg` to warning
```
Comment (korydraughn) [inline docs/release_notes.md:89]: Remove indented lines once L82 is confirmed to be good.

File: docs/release_notes.md [hunk 12]
```diff
@@ -1,5 +1,176 @@
...
+
+- Remove ACLs following removal of user (#7778).
+
+    [#7778] User removal now also removes ACLs
```
Comment (korydraughn) [inline docs/release_notes.md:93]: Remove indented lines once L91 is confirmed to be good.

File: docs/release_notes.md [hunk 13]
```diff
@@ -1,5 +1,176 @@
...
+
+- Migrate GitHub workflows to Ubuntu 24.04 (#7823).
+
+- Expand permission levels supported by `rc_atomic_apply_acl_operations` API (#7913).
```
Comment (korydraughn) [inline docs/release_notes.md:97]: Remove empty line.

File: docs/release_notes.md [hunk 14]
```diff
@@ -1,5 +1,176 @@
...
+
+    [#7913] Add support for same range of permissions ichmod allows
```
Comment (korydraughn) [inline docs/release_notes.md:99]: Remove indented lines once L97 is confirmed to be good.

File: docs/release_notes.md [hunk 15]
```diff
@@ -1,5 +1,176 @@
...
+
+- Log message only when oprType is changed to `UNREG_OPR` (#7953).
+
+    [#7953] Do not log message unless oprType is changed to UNREG_OPR.
```
Comment (korydraughn) [inline docs/release_notes.md:103]: Remove indented lines once L101 is confirmed to be good.

File: docs/release_notes.md [hunk 16]
```diff
@@ -1,5 +1,176 @@
...
+
+- Remove installation of packedRei directory from CMakeLists.txt (#7993).
+- Install msiExecCmd_bin/hello script as a template (#7994).
+- Print user-friendly error message in `itree` (#8082).
+
+- Improve GenQuery2 parser's parsing of whitespace sequences (#8182).
+
+    [#8182] Update GenQuery2 flex rule to match whitespace sequences.
```
Comment (korydraughn) [inline docs/release_notes.md:111]: Remove indented lines once L109 is confirmed to be good.

File: docs/release_notes.md [hunk 17]
```diff
@@ -1,5 +1,176 @@
...
+
+- Add json_events.hpp header file to development package (#8200).
+
+    [#8200] Add `json_events.hpp` to CMake
```
Comment (korydraughn) [inline docs/release_notes.md:115]: Remove indented lines once L113 is confirmed to be good.

File: docs/release_notes.md [hunk 18]
```diff
@@ -1,5 +1,176 @@
...
+Release Date: 2025-03-XX
+
+The iRODS Consortium and RENCI are pleased to announce iRODS 4.3.4.
+
+This release ...
+
+- deprecations in support of irods 5
+- ssl verification
+- bug fixes
+- clean up
```
Comment (korydraughn) [inline docs/release_notes.md:14]: Add summary.

File: docs/release_notes.md [hunk 19]
```diff
@@ -1,5 +1,176 @@
...
+
+TODO: Consolidate commits
```
Comment (korydraughn) [inline docs/release_notes.md:16]: Remove.

File: docs/release_notes.md [hunk 20]
```diff
@@ -1,5 +1,179 @@
...
+The iRODS Consortium and RENCI are pleased to announce iRODS 4.3.4.
+
+This release ...
+
+- deprecations in support of irods 5
+- ssl verification
+- bug fixes
+- clean up
+
+This release consists of [93 commits from 7 contributors](https://github.com/irods/irods/compare/4.3.3...4.3.4) and [closed 132 issues marked for 4.3.4](https://github.com/irods/irods/issues?q=milestone%3A4.3.4).
```
Comment (korydraughn) [inline docs/release_notes.md:16]: Consider changing 7 to 6. The only thing is that we provide a link to GitHub which reports 7.

From @trel:
```
$ git shortlog -ns 4.3.3...4-3-stable
    29  Kory Draughn
    23  Markus Kitsinger (SwooshyCueb)
    18  Derek D
    12  Alan King
     7  Martin Jaime Flores Jr.
     3  Martin Jaime Flores Jr
     1  Justin James
```

Comment (trel) [reply docs/release_notes.md:16]: I think there is precedent in using the correct number, in this case 6.    GitHub doesn't know all of Martin's secrets.

Comment (korydraughn) [reply docs/release_notes.md:16]: I agree. Plus, the output you provided from `git` is now part of this PR's history.

Will change it to 6.

File: docs/release_notes.md (hunk 5 above)
Comment (korydraughn) [reply docs/release_notes.md]: Done.

File: docs/release_notes.md (hunk 6 above)
Comment (korydraughn) [reply docs/release_notes.md]: Done.

File: docs/release_notes.md (hunk 7 above)
Comment (korydraughn) [reply docs/release_notes.md]: Done.

File: docs/release_notes.md (hunk 10 above)
Comment (korydraughn) [reply docs/release_notes.md]: Done.

File: docs/release_notes.md (hunk 11 above)
Comment (korydraughn) [reply docs/release_notes.md]: Done.

File: docs/release_notes.md (hunk 12 above)
Comment (korydraughn) [reply docs/release_notes.md]: Done.

File: docs/release_notes.md (hunk 13 above)
Comment (korydraughn) [reply docs/release_notes.md]: Done.

File: docs/release_notes.md (hunk 14 above)
Comment (korydraughn) [reply docs/release_notes.md]: Done.

File: docs/release_notes.md (hunk 15 above)
Comment (korydraughn) [reply docs/release_notes.md]: Done.

File: docs/release_notes.md (hunk 17 above)
Comment (korydraughn) [reply docs/release_notes.md]: Done.

File: docs/release_notes.md (hunk 19 above)
Comment (korydraughn) [reply docs/release_notes.md]: Done.

File: docs/release_notes.md (hunk 20 above)
Comment (korydraughn) [reply docs/release_notes.md]: Done.

File: docs/release_notes.md (hunk 18 above)
Comment (korydraughn) [reply docs/release_notes.md]: First draft done.

**Comment by korydraughn** [comment]
I've confirmed that all issue numbers are included in the milestone. There are a few which are attached to the ongoing milestone and 5.0.0 milestone. Haven't made a decision about those yet.

There's also one case where an issue number is referenced twice. Will think about that.

File: docs/release_notes.md [hunk 21]
```diff
@@ -1,5 +1,97 @@
...
+
+### Removed
+
+- Remove CentOS 7 build from GitHub workflows (#7968).
+
+### Deprecated
+
+- Deprecate `ilocate` (#2524).
+- Deprecate `igetwild` (#2525).
+- Deprecate modification of user having rodsgroup type in GeneralAdmin API (#2978).     THIS ISSUE IS REFERENCED IN TWO SECTIONS. IT DOES COVER TWO THINGS.
```
Comment (korydraughn) [inline docs/release_notes.md:44]: Normally, we want to reference an issue number once, but this feels like an exceptional case due to what the issue covers.

File: docs/release_notes.md [hunk 22]
```diff
@@ -1,5 +1,97 @@
...
+- Deprecate server monitoring microservices (#7977).
+- Deprecate DataObjLock and DataObjUnlock (#7979).
+- Deprecate unused global variables in resource manager implementation. (#8042).
+- Deprecate `fillGenQueryInpFromStrCond` (#8088).
+- Deprecate `forkAndExec` (#8109).
+
+### Fixed
+
+- Avoid SIGABRT by setting pointers to null after deallocation (#3581).
+- Return errors on one-sided encryption in parallel transfer (#4984).       THINK ABOUT THIS ONE. ISSUE IS OPEN, BUT HAS BEEN BUMPED TO 5.0 MILESTONE.
```
Comment (korydraughn) [inline docs/release_notes.md:59]: I'm wondering if issue 4984 should be closed instead of being moved into the 5.0.0 milestone.

Hmm.

File: docs/release_notes.md [hunk 23]
```diff
@@ -1,5 +1,97 @@
...
+- Return error on invalid GenQuery1 aggregate function (#8080).
+- Return error on nonexistent target for write functions in iRODS Rule Language (#8095).
+- Clean up memory in `rsDataObjChksum`, various endpoints, and iCommands (#8106).
+- Correct narrowing of floating point values (#8110).
+- Remove constructor from `Cache` data type in iRODS Rule Language to allow for `memset` (#8111).
+- Avoid segfault in `rcDisconnect` (#8120).
+- Fix libstdc++ linker error for `ienv`, `ierror`, and `ipwd` on EL8 (#8122).
+- Fix GenQuery2 column mappings for user zone (#8134).
+- Use correct GenQuery2 table aliases for permissions (#8135).
+- Fix `nullptr` to function argument and unsigned overflow (#8153).        THIS IS PART OF THE ONGOING MILESTONE.
```
Comment (korydraughn) [inline docs/release_notes.md:81]: This is likely okay since it's no different that housekeeping issue numbers landing in the release notes.

File: docs/release_notes.md (hunk 22 above)
Comment (trel) [reply docs/release_notes.md:59]: 'Changed' does sound a bit more correct to me.

It is definitely cleaner to have work done in one release be closed during that release... but there is precedent in us having commits in the system that do NOT yet close an issue.   However, I don't think we've ever put issue numbers in the release notes that weren't closed.   So... we're in new territory either way.

Smaller/discrete issues are the real answer.

File: docs/release_notes.md (hunk 21 above)
Comment (korydraughn) [reply docs/release_notes.md:44]: I get what you're saying, but if I remove one of them, we lose important information. The words for each section are important IMO.

If I were going to leave one in, it would probably be the deprecation message since it could be viewed as - _The Consortium must be providing a better alternative._ The line under the "Added" section doesn't give the reader any reason to think that the old way is possibly deprecated (assuming they even know about the old way).

Comment (trel) [reply docs/release_notes.md:44]: Understood.   I think in the long run ... as long as you have a sense for how you want to hold them... and can remain consistent... it's fine.

Comment (alanking) [reply docs/release_notes.md:44]: On that note, `iadmin mkuser` with a `rodsgroup` is also deprecated as of this release.

File: docs/release_notes.md (hunk 22 above)
Comment (korydraughn) [reply docs/release_notes.md:59]: I agree with both comments.

Will move it to the "Changed" section.

I think it's important that we mention the change because it is a change in behavior. I'm going to leave it in even though the release notes normally only reference closed issues. The issue just gives the reader a place to start understanding the story.

File: docs/release_notes.md (hunk 21 above)
Comment (korydraughn) [reply docs/release_notes.md:44]: Do you feel there's something we should tweak?

Comment (korydraughn) [reply docs/release_notes.md]: Updated to match the suggestion.

File: docs/release_notes.md (hunk 22 above)
Comment (korydraughn) [reply docs/release_notes.md]: Moved to "Changed" section.

File: docs/release_notes.md (hunk 23 above)
Comment (korydraughn) [reply docs/release_notes.md]: Done.

File: docs/release_notes.md [hunk 24]
```diff
@@ -1,5 +1,97 @@
...
 
+## 4.3.4
+
+Release Date: 2025-03-05
+
+The iRODS Consortium and RENCI are pleased to announce iRODS 4.3.4.
+
+This release primarily focuses on preparing the ground for the initial release of iRODS 5 by introducing more deprecations, fixing bugs, and cleaning up implementations.
+
+The biggest news of this release is that `imiscsvrinfo` has been updated to report SSL/TLS certificate information, if in use. This is made possible due to enhancements to the `rcGetMiscSvrInfo` API endpoint. Keep in mind that this requires the client to have a proper client-side configuration.
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply docs/release_notes.md]: Remove trailing space.

**Review by alanking — COMMENTED** [review]
Seems good to me. Squash when ready

**Comment by trel** [comment]
yes squash to two please.  can make the readme the last commit.

**Comment by korydraughn** [comment]
Squashed w/o pounds.

**Comment by trel** [comment]
i think it's ready - let's wait for alan.

**Review by alanking — APPROVED** [review]
\# it
